From a5a472bf6c627f666020846d13b023ca997dd301 Mon Sep 17 00:00:00 2001
From: John Greene <jogreene@redhat.com>
Date: Wed, 10 Sep 2014 18:28:36 -0400
Subject: [ethernet] ixgbe: Check register reads for adapter removal

Message-id: <1410373750-11341-17-git-send-email-jogreene@redhat.com>
Patchwork-id: 93546
O-Subject: [RHEL7.1 PATCH 16/50] ixgbe: Check register reads for adapter removal
Bugzilla: 1091123
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>

BZ: 1091123
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7919168
Tested: In beaker. See BZ for details

Check all register reads for adapter removal by checking the status
register after any register read that returns 0xFFFFFFFF. Since the
status register will never return 0xFFFFFFFF unless the adapter is
removed, such a value from a status register read confirms the
removal.

Signed-off-by: Mark Rustad <mark.d.rustad@intel.com>
Tested-by: Phil Schmitt <phillip.j.schmitt@intel.com>
Signed-off-by: Aaron Brown <aaron.f.brown@intel.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 2a1a091c1ff898b1de4dcac2236b85f88aa9d51f)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/ixgbe/ixgbe.h b/drivers/net/ethernet/intel/ixgbe/ixgbe.h
index 2d06a80..00e7e4c 100644
--- a/drivers/net/ethernet/intel/ixgbe/ixgbe.h
+++ b/drivers/net/ethernet/intel/ixgbe/ixgbe.h
@@ -731,6 +731,7 @@ struct ixgbe_adapter {
 #ifdef IXGBE_FCOE
  struct ixgbe_fcoe fcoe;
 #endif /* IXGBE_FCOE */
+ u8 __iomem *io_addr; /* Mainly for iounmap use */
  u32 wol;
 
  u16 bd_number;
diff --git a/drivers/net/ethernet/intel/ixgbe/ixgbe_common.h b/drivers/net/ethernet/intel/ixgbe/ixgbe_common.h
index 54ddc33..27a770e 100644
--- a/drivers/net/ethernet/intel/ixgbe/ixgbe_common.h
+++ b/drivers/net/ethernet/intel/ixgbe/ixgbe_common.h
@@ -127,6 +127,15 @@ void ixgbe_set_rxpba_generic(struct ixgbe_hw *hw, int num_pb,
 s32 ixgbe_get_thermal_sensor_data_generic(struct ixgbe_hw *hw);
 s32 ixgbe_init_thermal_sensor_thresh_generic(struct ixgbe_hw *hw);
 
+#define IXGBE_FAILED_READ_REG 0xffffffffU
+
+static inline bool ixgbe_removed(void __iomem *addr)
+{
+ return unlikely(!addr);
+}
+
+void ixgbe_check_remove(struct ixgbe_hw *hw, u32 reg);
+
 static inline void ixgbe_write_reg(struct ixgbe_hw *hw, u32 reg, u32 value)
 {
  writel(value, hw->hw_addr + reg);
@@ -150,7 +159,15 @@ static inline void ixgbe_write_reg64(struct ixgbe_hw *hw, u32 reg, u64 value)
 
 static inline u32 ixgbe_read_reg(struct ixgbe_hw *hw, u32 reg)
 {
- return readl(hw->hw_addr + reg);
+ u8 __iomem *reg_addr = ACCESS_ONCE(hw->hw_addr);
+ u32 value;
+
+ if (ixgbe_removed(reg_addr))
+  return IXGBE_FAILED_READ_REG;
+ value = readl(reg_addr + reg);
+ if (unlikely(value == IXGBE_FAILED_READ_REG))
+  ixgbe_check_remove(hw, reg);
+ return value;
 }
 #define IXGBE_READ_REG(a, reg) ixgbe_read_reg((a), (reg))
 
diff --git a/drivers/net/ethernet/intel/ixgbe/ixgbe_main.c b/drivers/net/ethernet/intel/ixgbe/ixgbe_main.c
index fb38ae2..deca82b 100644
--- a/drivers/net/ethernet/intel/ixgbe/ixgbe_main.c
+++ b/drivers/net/ethernet/intel/ixgbe/ixgbe_main.c
@@ -283,6 +283,35 @@ static void ixgbe_service_event_schedule(struct ixgbe_adapter *adapter)
   schedule_work(&adapter->service_task);
 }
 
+static void ixgbe_remove_adapter(struct ixgbe_hw *hw)
+{
+ struct ixgbe_adapter *adapter = hw->back;
+
+ if (!hw->hw_addr)
+  return;
+ hw->hw_addr = NULL;
+ e_dev_err("Adapter removed\n");
+}
+
+void ixgbe_check_remove(struct ixgbe_hw *hw, u32 reg)
+{
+ u32 value;
+
+ /* The following check not only optimizes a bit by not
+  * performing a read on the status register when the
+  * register just read was a status register read that
+  * returned IXGBE_FAILED_READ_REG. It also blocks any
+  * potential recursion.
+  */
+ if (reg == IXGBE_STATUS) {
+  ixgbe_remove_adapter(hw);
+  return;
+ }
+ value = ixgbe_read_reg(hw, IXGBE_STATUS);
+ if (value == IXGBE_FAILED_READ_REG)
+  ixgbe_remove_adapter(hw);
+}
+
 static void ixgbe_service_event_complete(struct ixgbe_adapter *adapter)
 {
  BUG_ON(!test_bit(__IXGBE_SERVICE_SCHED, &adapter->state));
@@ -2969,7 +2998,7 @@ void ixgbe_configure_tx_ring(struct ixgbe_adapter *adapter,
    ring->count * sizeof(union ixgbe_adv_tx_desc));
  IXGBE_WRITE_REG(hw, IXGBE_TDH(reg_idx), 0);
  IXGBE_WRITE_REG(hw, IXGBE_TDT(reg_idx), 0);
- ring->tail = hw->hw_addr + IXGBE_TDT(reg_idx);
+ ring->tail = adapter->io_addr + IXGBE_TDT(reg_idx);
 
  /*
   * set WTHRESH to encourage burst writeback, it should not be set
@@ -3372,7 +3401,7 @@ void ixgbe_configure_rx_ring(struct ixgbe_adapter *adapter,
    ring->count * sizeof(union ixgbe_adv_rx_desc));
  IXGBE_WRITE_REG(hw, IXGBE_RDH(reg_idx), 0);
  IXGBE_WRITE_REG(hw, IXGBE_RDT(reg_idx), 0);
- ring->tail = hw->hw_addr + IXGBE_RDT(reg_idx);
+ ring->tail = adapter->io_addr + IXGBE_RDT(reg_idx);
 
  ixgbe_configure_srrctl(adapter, ring);
  ixgbe_configure_rscctl(adapter, ring);
@@ -7538,6 +7567,7 @@ static int ixgbe_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
 
  hw->hw_addr = ioremap(pci_resource_start(pdev, 0),
          pci_resource_len(pdev, 0));
+ adapter->io_addr = hw->hw_addr;
  if (!hw->hw_addr) {
   err = -EIO;
   goto err_ioremap;
@@ -7856,7 +7886,7 @@ err_register:
 err_sw_init:
  ixgbe_disable_sriov(adapter);
  adapter->flags2 &= ~IXGBE_FLAG2_SEARCH_FOR_SFP;
- iounmap(hw->hw_addr);
+ iounmap(adapter->io_addr);
 err_ioremap:
  free_netdev(netdev);
 err_alloc_etherdev:
@@ -7923,7 +7953,7 @@ static void ixgbe_remove(struct pci_dev *pdev)
  kfree(adapter->ixgbe_ieee_ets);
 
 #endif
- iounmap(adapter->hw.hw_addr);
+ iounmap(adapter->io_addr);
  pci_release_selected_regions(pdev, pci_select_bars(pdev,
          IORESOURCE_MEM));
 
-- 
1.7.1